dandischema/digests/zarr.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import hashlib
+from typing import Dict, List, Optional
+
+import pydantic
+
+
+"""Passed to the json() method of pydantic models for serialization."""
+ENCODING_KWARGS = {"separators": (",", ":")}
+
+
+class ZarrChecksum(pydantic.BaseModel):
+    """
+    A checksum for a single file/directory in a zarr file.
+
+    Every file and directory in a zarr archive has a path and a MD5 hash.
+    """
+
+    md5: str
+    path: str
+
+    # To make ZarrChecksums sortable
+    def __lt__(self, other: ZarrChecksum):
+        return self.path < other.path
+
+
+class ZarrChecksums(pydantic.BaseModel):
+    """
+    A set of file and directory checksums.
+
+    This is the data hashed to calculate the checksum of a directory.
+    """
+
+    directories: List[ZarrChecksum] = pydantic.Field(default_factory=list)
+    files: List[ZarrChecksum] = pydantic.Field(default_factory=list)
+
+    @property
+    def is_empty(self):
+        return self.files == [] and self.directories == []
+
+    def _index(self, checksums: List[ZarrChecksum], checksum: ZarrChecksum):
```

Comment:
I suggest you consider using the [`bisect`](https://docs.python.org/3/library/bisect.html) module.